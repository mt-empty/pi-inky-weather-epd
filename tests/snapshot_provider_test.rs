//! Provider-specific snapshot tests
//!
//! These tests verify that the complete dashboard generation pipeline produces
//! consistent output for each weather provider (BOM and Open-Meteo).
//!
//! ## How These Tests Work
//!
//! 1. **Fixed Time**: Use FixedClock to ensure deterministic "current hour"
//! 2. **Fixture Data**: Load pre-defined weather data (no API calls)
//! 3. **Full Pipeline**: Run complete dashboard generation
//! 4. **Snapshot SVG**: Capture and compare the full SVG output
//!
//! ## Running These Tests
//!
//! **Default (Open-Meteo)**:
//! ```bash
//! TEST_MODE=config/test cargo test --test snapshot_provider_test
//! ```
//!
//! **BOM Provider** (requires override):
//! ```bash
//! TEST_MODE=config/test APP_API__PROVIDER=bom cargo test --test snapshot_provider_test snapshot_bom_dashboard -- --ignored
//! ```
//!
//! **Important**: Set `TEST_MODE=config/test` to load test configuration.
//! This loads `config/test.toml` which sets:
//! - `provider = "open_meteo"` (default test provider)
//! - `disable_weather_api_requests = true` (use fixtures, no real API calls)
//! - `weather_data_cache_path = "tests/fixtures/"` (load from test fixtures)
//!
//! ## Reviewing Snapshots
//!
//! On first run or after intentional changes:
//! ```bash
//! TEST_MODE=config/test cargo test --test snapshot_provider_test
//! cargo insta review  # Review and accept/reject changes
//! ```

use pi_inky_weather_epd::{clock::FixedClock, generate_weather_dashboard_with_clock, CONFIG};
use std::fs;

/// Test Open-Meteo provider dashboard generation with fixed time and fixtures
///
/// **Fixed Time**: Oct 10, 2025, 1:00 AM UTC = Oct 10, 2025, 12:00 PM Melbourne (AEDT)
///
/// **Fixture Files Used**:
/// - `tests/fixtures/open_meteo_forecast.json` - Combined hourly + daily data
///
/// **What This Tests**:
/// - Open-Meteo API response parsing (parallel array structure)
/// - Float temperature handling
/// - Global provider (lat/lon instead of geohash)
/// - GMT timezone conversion to Melbourne (AEDT)
/// - Dashboard rendering with Open-Meteo data structure
///
/// **Snapshot Captures**:
/// - Complete SVG structure (template + rendered data)
/// - Current weather display (temp, icon, feels like, wind, humidity, UV)
/// - Hourly forecast graph (temperature, rain)
/// - 7-day forecast cards (only 3 days in fixture)
/// - Sunrise/sunset times
///
/// **NOTE**: This is the default test provider (config/test.toml sets provider = "open_meteo").
/// No special environment variables needed.
#[test]
fn snapshot_open_meteo_dashboard() {
    // Verify test configuration is loaded
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test requires TEST_MODE=config/test environment variable"
    );
    assert_eq!(
        CONFIG.misc.weather_data_cache_path.to_str().unwrap(),
        "tests/fixtures/",
        "Test config should use fixtures path"
    );
    assert_eq!(
        format!("{}", CONFIG.api.provider).to_lowercase(),
        "openmeteo",
        "Test config should use Open-Meteo provider"
    );

    // Fixed time: Oct 10, 2025, 1:00 AM UTC
    // In Melbourne AEDT (UTC+11): Oct 10, 2025, 12:00 PM (noon)
    // This should match mid-point of hourly forecast data
    let clock =
        FixedClock::from_rfc3339("2025-10-10T01:00:00Z").expect("Failed to create fixed clock");

    // Verify test configuration is loaded
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test config should have disable_weather_api_requests = true"
    );

    // Temporarily override provider to Open-Meteo
    // Note: This requires the config to be reloadable or we need a different approach
    // For now, we'll test with whatever provider is set in test.toml

    // Generate the dashboard with Open-Meteo provider
    let result = generate_weather_dashboard_with_clock(&clock);
    assert!(
        result.is_ok(),
        "Failed to generate Open-Meteo dashboard: {:?}",
        result.err()
    );

    // Read the generated SVG
    let svg_content = fs::read_to_string(CONFIG.misc.generated_svg_name.clone())
        .expect("Failed to read generated SVG file");

    // Verify SVG is not empty
    assert!(!svg_content.is_empty(), "Generated SVG should not be empty");
    assert!(
        svg_content.contains("<svg"),
        "Generated file should be valid SVG"
    );

    // Snapshot the SVG structure
    // This captures the complete dashboard including:
    // - Weather data from Open-Meteo fixtures
    // - Rendered at fixed time (12:00 PM Melbourne)
    // - All graph paths, labels, icons
    insta::assert_snapshot!("open_meteo_dashboard", svg_content);
}

/// Test BOM provider dashboard generation with fixed time and fixtures
///
/// **Prerequisites**: Run with `TEST_MODE=config/test` environment variable
///
/// **Fixed Time**: Oct 9, 2025, 10:00 PM UTC = Oct 10, 2025, 9:00 AM Melbourne (AEDT)
///
/// **Fixture Files Used**:
/// - `tests/fixtures/bom_hourly_forecast.json` - 4 hours of data
/// - `tests/fixtures/bom_daily_forecast.json` - 3 days of data
///
/// **What This Tests**:
/// - BOM-specific API response parsing
/// - Integer temperature handling
/// - Knots to km/h wind conversion
/// - Australian timezone (AEDT) display
/// - Dashboard rendering with BOM data structure
///
/// **Snapshot Captures**:
/// - Complete SVG structure (template + rendered data)
/// - Current weather display (temp, icon, feels like, wind, humidity, UV)
/// - Hourly forecast graph (temperature, rain)
/// - 7-day forecast cards
/// - Sunrise/sunset times
///
/// **NOTE**: This test requires APP_API__PROVIDER environment variable override:
/// ```bash
/// TEST_MODE=config/test APP_API__PROVIDER=bom cargo test --test snapshot_provider_test snapshot_bom_dashboard -- --ignored
/// cargo insta review
/// ```
/// Note: Double underscore __ is used as separator for nested config keys (api.provider -> API__PROVIDER)
#[test]
#[ignore] // Run separately with APP_API__PROVIDER=bom (CONFIG initializes once per test binary)
fn snapshot_bom_dashboard() {
    // Verify test configuration is loaded (fail fast if TEST_MODE not set)
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test requires TEST_MODE=config/test environment variable"
    );
    assert_eq!(
        CONFIG.misc.weather_data_cache_path.to_str().unwrap(),
        "tests/fixtures/",
        "Test config should use fixtures path"
    );
    assert_eq!(
        format!("{}", CONFIG.api.provider).to_lowercase(),
        "bom",
        "Test should use BOM provider (set APP_API__PROVIDER=bom)"
    );

    // Fixed time: Oct 9, 2025, 10:00 PM UTC
    // In Melbourne AEDT (UTC+11): Oct 10, 2025, 9:00 AM
    // This matches the first hourly forecast data point in the fixture
    let clock =
        FixedClock::from_rfc3339("2025-10-09T22:00:00Z").expect("Failed to create fixed clock");

    // Generate the dashboard with BOM provider
    let result = generate_weather_dashboard_with_clock(&clock);
    assert!(
        result.is_ok(),
        "Failed to generate BOM dashboard: {:?}",
        result.err()
    );

    // Read the generated SVG
    let svg_content = fs::read_to_string(CONFIG.misc.generated_svg_name.clone())
        .expect("Failed to read generated SVG file");

    // Verify SVG is not empty
    assert!(!svg_content.is_empty(), "Generated SVG should not be empty");
    assert!(
        svg_content.contains("<svg"),
        "Generated file should be valid SVG"
    );

    // Snapshot the SVG structure
    // This captures the complete dashboard including:
    // - Weather data from BOM fixtures
    // - Rendered at fixed time (9:00 AM Melbourne)
    // - All graph paths, labels, icons
    insta::assert_snapshot!("bom_dashboard", svg_content);
}
