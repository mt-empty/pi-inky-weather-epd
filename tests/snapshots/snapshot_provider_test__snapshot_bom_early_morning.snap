---
source: tests/snapshot_provider_test.rs
expression: svg_content
---
<svg width="800" height="480" font-family="Roboto, sans-serif" xmlns="http://www.w3.org/2000/svg">

    <!-- tspan doesn't work well with the degree symbol ° for some fonts!-->
    <!-- Avoid using tspan with text-anchor, `dx` or `dy`, resvg doesn't handle it properly -->
    <!-- see https://github.com/linebender/resvg/issues/583 -->

    <defs>
        <linearGradient id="fillGradient" x1="0%" y1="100%" x2="0%" y2="0%">
            <stop offset="0%" stop-color="blue" stop-opacity="0.5" />
            <stop offset="100%" stop-color="blue" stop-opacity="0.1" />
        </linearGradient>
        <pattern id="dots" patternUnits="userSpaceOnUse" width="10" height="10">
            <circle cx="5" cy="5" r="2" fill="blue" />
        </pattern>
        <pattern id="grid" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 10 0 L 0 0 0 10" fill="none" stroke="blue" stroke-width="1" />
        </pattern>
        <pattern id="diagonal-lines" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 0 10 L 10 0" stroke="blue" stroke-width="2" />
        </pattern>
        <pattern id="crosshatch" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 0 0 L 10 10 M 10 0 L 0 10" stroke="blue" stroke-width="1" />
        </pattern>
    </defs>

    <rect width="100%" height="100%" fill="white" />

    <!-- Current Day Information -->
    <!-- Due to resvg bug, the position is intentionally off to compensate for the bug -->
    <!-- Do not modify the x position of current_hour_temp and current_hour_feels_like, see above issue at the start of the file -->
    <text x="270" y="50" font-size="35" fill="black">Sunday, 26 October </text>
    <image x="0" y="0" width="200" height="180" href="static/fill-svg-static/overcast-night-rain.svg" />


    <!-- Current temperature and Feels Like temperature -->
    <text x="266" y="128" font-size="70" fill="red" text-anchor="middle">
        14
        <tspan baseline-shift="27" dx="-20" font-size="33">°C</tspan>
    </text>

    <text x="246" y="158" text-anchor="middle" font-size="18" fill="black">
        <tspan>Feels</tspan>
        <tspan dy="15.5" dx="-46.6">Like</tspan>
        <tspan font-family="Roboto-Regular-Dashed" dominant-baseline="middle" font-size="55" fill="green"
            dx="10" dy="-7">
            12
            <tspan baseline-shift="12" dx="-15" font-size="30">°</tspan>
        </tspan>
    </text>

    <!-- Diagnostic Message -->
    <svg x="580" y="0" width="200" height="100" text-anchor="middle" visibility="hidden">
        <!-- Cascading diagnostic icons (stacked by priority) -->
        
        <!-- Message for highest priority error only -->
        <text x="100" y="60" width="200" font-size="12">
            NA
        </text>
    </svg>

    <!-- Sunset/Sunrise Information -->
    <svg x="30" y="150">
        <image x="0" y="0" width="75" height="75" href="static/fill-svg-static/sunrise.svg" />
        <text x="12" y="70" text-anchor="start" font-size="20" fill="black">06:21</text>

        <image x="70" y="0" width="75" height="75" href="static/fill-svg-static/sunset.svg" />
        <text x="83" y="70" text-anchor="start" font-size="20" fill="black">19:48</text>
    </svg>

    <g transform="translate(-15, 0)">

        <!-- vertical lines with artistic variations -->
        <svg class="line">
            <path d="M95 240 C90 220, 95 320, 95 470" />
            <path d="M150 240 C155 220, 145 320, 150 470" />

            <!-- horizontal lines with artistic variations -->
            <path d="M40 260 C150 261, 250 258, 185 260" />
            <path d="M40 330 C150 325, 250 335, 185 330" />
            <path d="M40 400 C150 402, 250 399, 185 400" />
        </svg>

        <!-- row header-->
        <svg x="15" y="210" label="UV-index" text-anchor="start">
            <text x="16" y="43.5" fill="black" font-size="20">Metric
            </text>

            <text x="85" y="43.5" fill="black" font-size="20">Now
            </text>
            <text x="140" y="43.5" fill="black" font-size="20">Max
                <tspan baseline-shift="super" font-size="17" dx="-3">24h</tspan>
            </text>
        </svg>

        <!-- UV Index -->
        <svg x="20" y="260" label="UV-index" text-anchor="middle">
            <image width="75" height="75" href="static/fill-svg-static/uv-index-none.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">0
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="normal">6
            </text>
        </svg>

        <!-- Wind Information -->
        <svg x="20" y="330" label="wind" text-anchor="middle">
            <text x="42.2" y="66.2" fill="black" font-size="15.1" text-anchor="middle">
                km/h
            </text>
            <image width="75" height="75" href="static/fill-svg-static/wind.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">7
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="italic">26
            </text>
        </svg>

        <!-- Relative Humidity -->
        <svg x="20" y="400" label="relative_humidity" text-anchor="middle">
            <image width="75" height="75" href="static/fill-svg-static/humidity-plus-plus.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">85
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="italic">85
            </text>
        </svg>
    </g>


    <!-- Forecast for the next 6 days -->
    <svg x="292" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Mon</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/overcast-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">15°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">11°</tspan>
        </text>
    </svg>

    <svg x="373" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Tue</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">19°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">6°</tspan>
        </text>
    </svg>

    <svg x="454" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Wed</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">21°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">8°</tspan>
        </text>
    </svg>

    <svg x="535" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Thu</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">23°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">11°</tspan>
        </text>
    </svg>

    <svg x="616" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Fri</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/overcast-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">27°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">15°</tspan>
        </text>
    </svg>

    <svg x="697" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Sat</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/partly-cloudy-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">25°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">12°</tspan>
        </text>
    </svg>

    <!-- Graph for hourly forecast -->
    <svg x="220" y="160" width="560" height="350" viewBox="-50 0 690 300" preserveAspectRatio="xMidYMid meet"
        style="overflow: visible;">
        <path stroke="black" stroke-linejoin="round" stroke-width="2" d="M 0 300 L 600 300 M 0 295 L 0 305 M 100 295 L 100 305 M 200 295 L 200 305 M 300 295 L 300 305 M 400 295 L 400 305 M 500 295 L 500 305 M 600 295 L 600 305" fill="none" />
        <!-- X axis Guidelines -->
        <path stroke="black" stroke-linejoin="round" stroke-width="1" d="M 0 300 L 600 300 M 0 300 v -300 m 0 2 v -2 M 100 300 v -300 m 0 2 v -2 M 200 300 v -300 m 0 2 v -2 M 300 300 v -300 m 0 2 v -2 M 400 300 v -300 m 0 2 v -2 M 500 300 v -300 m 0 2 v -2 M 600 300 v -300 m 0 2 v -2" fill="none"
            stroke-dasharray="5,10" />
        <!-- X Labels -->
        <text x="0" y="320" fill="black" font-size="19" text-anchor="middle">6am</text><text x="100" y="320" fill="black" font-size="19" text-anchor="middle">10am</text><text x="200" y="320" fill="black" font-size="19" text-anchor="middle">2pm</text><text x="300" y="320" fill="black" font-size="19" text-anchor="middle">6pm</text><text x="400" y="320" fill="black" font-size="19" text-anchor="middle">10pm</text><text x="500" y="320" fill="black" font-size="19" text-anchor="middle">2am</text><text x="600" y="320" fill="black" font-size="19" text-anchor="middle">6am</text><line x1="450" y1="0" x2="450" y2="300" stroke="black" stroke-width="2" stroke-dasharray="3,3" />
                   <text x="460" y="170" fill="black" font-size="19" font-style="italic"  transform="rotate(-90, 430, 135)" text-anchor="start">Monday</text>
        <!-- Y Labels -->
        <text x="-10" y="300"  fill="black" font-size="35" text-anchor="end" dx="8" dy="4">6°</text><text x="-10" y="240"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">9.0°</text><text x="-10" y="180"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">12.0°</text><text x="-10" y="120"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">15.0°</text><text x="-10" y="60"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">18.0°</text><text x="-10" y="0"  fill="black" font-size="35" text-anchor="end" dx="8" dy="4">21°</text>
        <!-- Y right Labels -->
        <text x="610" y="300" fill="black"  font-size="19" text-anchor="start" dy="4">0%</text><text x="610" y="240" fill="black"  font-size="19" text-anchor="start" dy="4">20%</text><text x="610" y="180" fill="black"  font-size="19" text-anchor="start" dy="4">40%</text><text x="610" y="120" fill="black"  font-size="19" text-anchor="start" dy="4">60%</text><text x="610" y="60" fill="black"  font-size="19" text-anchor="start" dy="4">80%</text><text x="610" y="0" fill="black"  font-size="19" text-anchor="start" dy="4">100%</text>
        <path stroke="red" stroke-linejoin="round" transform="translate(0, 300) scale(1, -1)"
            d="M 0.0000 160.0000C 34.7826 163.3333, 43.4783 170.0000, 52.1739 180.0000C 60.8696 190.0000, 69.5652 210.0000, 78.2609 220.0000C 86.9565 230.0000, 95.6522 233.3333, 104.3478 240.0000C 113.0435 246.6667, 121.7391 256.6667, 130.4348 260.0000C 139.1304 263.3333, 147.8261 260.0000, 156.5217 260.0000C 165.2174 260.0000, 173.9130 256.6667, 182.6087 260.0000C 191.3043 263.3333, 200.0000 273.3333, 208.6956 280.0000C 217.3913 286.6667, 226.0870 296.6667, 234.7826 300.0000C 243.4783 303.3333, 252.1739 303.3333, 260.8696 300.0000C 269.5652 296.6667, 278.2609 286.6667, 286.9565 280.0000C 295.6521 273.3333, 304.3478 266.6667, 313.0435 260.0000C 321.7391 253.3333, 330.4348 246.6667, 339.1304 240.0000C 347.8261 233.3333, 356.5217 226.6667, 365.2174 220.0000C 373.9131 213.3333, 382.6087 206.6667, 391.3044 200.0000C 400.0000 193.3333, 408.6956 186.6667, 417.3913 180.0000C 426.0869 173.3333, 434.7826 163.3333, 443.4782 160.0000C 452.1739 156.6667, 460.8696 163.3333, 469.5652 160.0000C 478.2609 156.6667, 486.9565 146.6667, 495.6522 140.0000C 504.3478 133.3333, 513.0435 123.3333, 521.7391 120.0000C 530.4348 116.6667, 539.1304 123.3333, 547.8260 120.0000C 556.5217 116.6667, 565.2173 103.3333, 573.9130 100.0000C 582.6086 96.6667, 595.6522 100.0000, 600.0000 100.0000" stroke-width="3.0" fill="none" />
        <path stroke="green" stroke-linejoin="round" stroke-dasharray="5,5"
            transform="translate(0, 300) scale(1, -1)" d="M 0.0000 123.3333C 34.7826 143.3333, 43.4783 133.3333, 52.1739 140.0000C 60.8696 146.6667, 69.5652 170.0000, 78.2609 180.0000C 86.9565 190.0000, 95.6522 193.3333, 104.3478 200.0000C 113.0435 206.6667, 121.7391 213.3333, 130.4348 220.0000C 139.1304 226.6667, 147.8261 236.6667, 156.5217 240.0000C 165.2174 243.3333, 173.9130 236.6667, 182.6087 240.0000C 191.3043 243.3333, 200.0000 256.6667, 208.6956 260.0000C 217.3913 263.3333, 226.0870 260.0000, 234.7826 260.0000C 243.4783 260.0000, 252.1739 260.0000, 260.8696 260.0000C 269.5652 260.0000, 278.2609 266.6667, 286.9565 260.0000C 295.6521 253.3333, 304.3478 233.3333, 313.0435 220.0000C 321.7391 206.6667, 330.4348 186.6667, 339.1304 180.0000C 347.8261 173.3333, 356.5217 183.3333, 365.2174 180.0000C 373.9131 176.6667, 382.6087 166.6667, 391.3044 160.0000C 400.0000 153.3333, 408.6956 146.6667, 417.3913 140.0000C 426.0869 133.3333, 434.7826 130.0000, 443.4782 120.0000C 452.1739 110.0000, 460.8696 90.0000, 469.5652 80.0000C 478.2609 70.0000, 486.9565 66.6667, 495.6522 60.0000C 504.3478 53.3333, 513.0435 46.6667, 521.7391 40.0000C 530.4348 33.3333, 539.1304 26.6667, 547.8260 20.0000C 556.5217 13.3333, 565.2173 3.3333, 573.9130 0.0000C 582.6086 -3.3333, 595.6522 0.0000, 600.0000 0.0000" stroke-width="3.0" fill="none" />
        <path transform="translate(0, 300) scale(1, -1)" d="M 0.0000 180.0000L 26.086956 180L 52.173912 210L 78.260864 210L 104.347824 210L 130.43478 180L 156.52173 180L 182.60869 180L 208.69565 210L 234.78261 210L 260.86957 210L 286.9565 210L 313.04346 210L 339.13043 210L 365.21738 120L 391.30435 120L 417.3913 120L 443.47824 120L 469.56522 120L 495.65216 120L 521.73914 150L 547.82605 150L 573.913 150L 600 120 L 600 0 L 0 0Z" fill="blue"
            fill-opacity="25%" />
        <defs>
            <linearGradient id="UVGradient" gradientUnits="objectBoundingBox" x1="0%" y1="0%" x2="100%" y2="0%">
                <stop offset="0.00%" stop-color="white"/><stop offset="4.35%" stop-color="white"/><stop offset="8.70%" stop-color="green"/><stop offset="13.04%" stop-color="green"/><stop offset="17.39%" stop-color="green"/><stop offset="21.74%" stop-color="yellow"/><stop offset="26.09%" stop-color="yellow"/><stop offset="30.43%" stop-color="yellow"/><stop offset="34.78%" stop-color="orange"/><stop offset="39.13%" stop-color="orange"/><stop offset="43.48%" stop-color="orange"/><stop offset="47.83%" stop-color="green"/><stop offset="52.17%" stop-color="green"/><stop offset="56.52%" stop-color="green"/><stop offset="60.87%" stop-color="white"/><stop offset="65.22%" stop-color="white"/><stop offset="69.57%" stop-color="white"/><stop offset="73.91%" stop-color="white"/><stop offset="78.26%" stop-color="white"/><stop offset="82.61%" stop-color="white"/><stop offset="86.96%" stop-color="white"/><stop offset="91.30%" stop-color="white"/><stop offset="95.65%" stop-color="white"/><stop offset="100.00%" stop-color="white"/>
            </linearGradient>
        </defs>
        <rect x="0" y="-14" width="600" height="8" fill="url(#UVGradient)" />
        <path stroke="black" stroke-linejoin="round" stroke-width="2" d="M 0 300 L 600 300 M 0 295 L 0 305 M 100 295 L 100 305 M 200 295 L 200 305 M 300 295 L 300 305 M 400 295 L 400 305 M 500 295 L 500 305 M 600 295 L 600 305" fill="none" />
        <path stroke="red" stroke-linejoin="round" stroke-width="2" d="M 0 0 L 0 300 M -5 300 L 5 300 M -5 240 L 5 240 M -5 180 L 5 180 M -5 120 L 5 120 M -5 60 L 5 60 M -5 0 L 5 0" />
        <path stroke="blue" stroke-linejoin="round" stroke-width="3" d="M 600 0 L 600 300 M 595 300 L 605 300 M 595 240 L 605 240 M 595 180 L 605 180 M 595 120 L 605 120 M 595 60 L 605 60 M 595 0 L 605 0"
            fill="none" />
    </svg>
    <!-- These are needed render svg in browser/ide -->
    <style>
        @font-face {
            font-family: 'Roboto';
            src: url('static/fonts/Roboto-Regular.ttf') format('truetype');
        }

        @font-face {
            font-family: 'Roboto-Regular-Dashed';
            src: url('static/fonts/Roboto-Regular-Dashed.ttf') format('truetype');
        }

        .line {
            stroke: black;
            stroke-width: 2;
            stroke-linecap: round;
            stroke-linejoin: round;
        }
    </style>
</svg>
